vault/plugin_catalog.go

```diff
@@ -0,0 +1,169 @@
+package vault
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"path/filepath"
+	"sort"
+	"strings"
+	"sync"
+
+	"github.com/hashicorp/vault/helper/builtinplugins"
+	"github.com/hashicorp/vault/helper/jsonutil"
+	"github.com/hashicorp/vault/helper/pluginutil"
+	"github.com/hashicorp/vault/logical"
+)
+
+var (
+	pluginCatalogPrefix = "plugin-catalog/"
+)
+
+// PluginCatalog keeps a record of plugins known to vault. External plugins need
+// to be registered to the catalog before they can be used in backends. Builtin
+// plugins are automatically detected and included in the catalog.
+type PluginCatalog struct {
+	catalogView *BarrierView
+	directory   string
+
+	lock sync.RWMutex
+}
+
+func (c *Core) setupPluginCatalog() error {
+	c.pluginCatalog = &PluginCatalog{
+		catalogView: c.systemBarrierView.SubView(pluginCatalogPrefix),
```

Comment:
I moved this over to `core/` it now creates a `NewBarrierView` with the path being `core/plugin-catalog`